File: Hashing/universal_hashing.py
# File Name: universal_hashing.py

TABLE_SIZE = 10
P = 101   # prime number

a = 7
b = 3

def hash_function(key):
    return ((a * key + b) % P) % TABLE_SIZE

def main():
    keys = [23, 45, 12, 67, 34]

    print("Universal Hashing Method")
    print("Table Size =", TABLE_SIZE, "\n")

    for key in keys:
        index = hash_function(key)
        print(f"Key: {key} -> Hash Index: {index}")

if __name__ == "__main__":
    main()

File: Hashing/universal_hashing.c
/* File Name: universal_hashing.c */

#include <stdio.h>

#define TABLE_SIZE 10
#define P 101   // prime number greater than max key

int a = 7;      // random constant (1 ≤ a ≤ p-1)
int b = 3;      // random constant (0 ≤ b ≤ p-1)

int hashFunction(int key) {
    return ((a * key + b) % P) % TABLE_SIZE;
}

int main() {
    int keys[] = {23, 45, 12, 67, 34};
    int n = sizeof(keys) / sizeof(keys[0]);

    printf("Universal Hashing Method\n");
    printf("Table Size = %d\n\n", TABLE_SIZE);

    for(int i = 0; i < n; i++) {
        int index = hashFunction(keys[i]);
        printf("Key: %d -> Hash Index: %d\n", keys[i], index);
    }

    return 0;
}
